src/main/kotlin/com/terraformation/backend/tracking/model/BiomassDetailsModel.kt

```diff
@@ -0,0 +1,108 @@
+import java.math.BigDecimal
+import java.time.Instant
+
+data class BiomassAdditionalSpeciesModel(
+    val isInvasive: Boolean,
+    val isThreatened: Boolean,
+    val speciesId: SpeciesId? = null,
+    val speciesName: String? = null,
+)
+
+data class BiomassQuadratSpeciesModel(
+    val abundancePercent: BigDecimal,
+    val isInvasive: Boolean,
+    val isThreatened: Boolean,
+    val speciesId: SpeciesId? = null,
+    val speciesName: String? = null,
+)
+
+data class BiomassQuadratModel(
+    val description: String? = null,
+    val species: List<BiomassQuadratSpeciesModel>,
+)
+
+data class RecordedBranchModel<ID : RecordedBranchId?>(
+    val id: ID,
+    val branchNumber: Int,
+    val description: String?,
+    val diameterAtBreastHeightCm: BigDecimal,
+    val isDead: Boolean,
+    val pointOfMeasurementM: BigDecimal,
+)
+
+typealias ExistingRecordedBranchModel = RecordedBranchModel<RecordedBranchId>
+
+typealias NewRecordedBranchModel = RecordedBranchModel<Nothing?>
+
+data class RecordedTreeModel<TreeId : RecordedTreeId?, BranchId : RecordedBranchId?>(
+    val id: TreeId,
+    val branches: List<RecordedBranchModel<BranchId>> = emptyList(),
+    val description: String? = null,
+    val diameterAtBreastHeightCm: BigDecimal? = null,
+    val heightM: BigDecimal? = null,
+    val isDead: Boolean,
+    val isTrunk: Boolean? = null,
+    val pointOfMeasurementM: BigDecimal? = null,
+    val shrubDiameterCm: BigDecimal? = null,
+    val speciesId: SpeciesId? = null,
+    val speciesName: String? = null,
+    val treeGrowthForm: TreeGrowthForm,
+    val treeNumber: Int,
+) {
+  fun validate() {
+    if (isTrunk == true && branches.isEmpty()) {
+      throw IllegalStateException("Tree trunk must contain at least one branch.")
+    }
+
+    if (isTrunk != true && branches.isNotEmpty()) {
+      throw IllegalStateException("Only tree trunk can contain branches.")
+    }
+  }
+}
+
+typealias ExistingRecordedTreeModel = RecordedTreeModel<RecordedTreeId, RecordedBranchId>
+
+typealias NewRecordedTreeModel = RecordedTreeModel<Nothing?, Nothing?>
+
+data class BiomassDetailsModel<
+    ID : ObservationId?,
+    PlotId : MonitoringPlotId?,
+    TreeId : RecordedTreeId?,
+    BranchId : RecordedBranchId?>(
+    val additionalSpecies: List<BiomassAdditionalSpeciesModel>,
+    val description: String? = null,
+    val forestType: BiomassForestType,
+    val herbaceousCoverPercent: BigDecimal,
+    val observationId: ID,
+    val ph: BigDecimal? = null,
+    val quadrats: Map<ObservationPlotPosition, BiomassQuadratModel>,
+    val salinityPpt: BigDecimal? = null,
+    val smallTreeCountRange: Pair<Int, Int>,
+    val soilAssessment: String,
+    val plotId: PlotId,
+    val tide: MangroveTide? = null,
+    val tideTime: Instant? = null,
+    val trees: List<RecordedTreeModel<TreeId, BranchId>>,
+    val waterDepthCm: BigDecimal? = null,
+) {
+  fun validate() {
+    // TODO convert database validation errors to human-readable error messages
```

Comment:
There's an argument to be made for a single source of truth for validation rules, but I wonder if trying to reverse-engineer database constraint failures would end up being more complex and/or brittle than doing validation in the application code (aside from foreign key constraints) where it's trivial to generate precise error messages, and treating the database rules as protection against bugs in application-level validation.